Thread safe queu

Fildo7525 opened the pull request:
Opravil som tam ten problem s dead lockom. Priority queue teraz funguje ako ma. Implementoval som si ju cez inheritance aby som vedel printnut cely jej obsah. Je to tiez thread safe.

Commit:
odometry: failes: get speeds

The robot does not send back the desired json string.
The response is that he received it.

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: The first response is if the request was received

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: changed logging levels

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: comments

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: be sure to create the files

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: show if the client connected

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: QUESTION: get the speeds of the motors

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: calculate the robot location

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: calculation the robot position

The calculation calculates the x, y coordinates of the robot
and its angle of rotation.

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: change debugging

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: measure the send and receive time

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
cp: reorganizing the includes

The includes are in this order:
    1. The header file
    2. Files from this project
    3. System files
    4. ROS2 files
    5. Cpp STL files

Includes are ordered in alphabetical order.

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
deprication: do not use getStoppedTimes

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
logs: Created merged log file from files of your choosing

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
README: updated

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
doxygen: include private methods to the documentation

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: add comments

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: limit the received velocities

When the robot does not move. The received messages contain excessively
massive numbers. The maximal angular speed that can be created on the
motor's shaft is limited by the software to 1000 rps.

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: the position is calculated using doubles

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: TODO

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: stopwatch: use stopwatch on every build type

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: measure the whole preparation code

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: synchronization on send and receive

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: comments

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: typo fix

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
global: reorganize private variables

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
global: fix some log statements

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: display changed velocities on change

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: create logs in separate directories

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: send int's

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: initialize the last stamp of velocity as a maximal value

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
client: display the message instead of the nnumber of sent bytes

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
logs: backup them before delete

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
FATAL: odometry: change the robot speed

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Fildo7525 committed:
odometry: get the wheel speed from const class

Co-authored-by: MichalDobis <miso.dobis@gmail.com>

Commit:
stopwatch: The functionality is needed int the release verion, too

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
queue: thread safe queue

To control the execution order of the send operations we created the
thead safe queue. It is ment to control the middle layer of client
operation.

To get the most accurate coordination we implemeted a priority queue
that prioritizes the speed requests that we push to it.

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
gitignore: ignore backup logs

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: log to collective file and be able to switch it off

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: log thread ID

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
client: set time limit on send and receive

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
client: the evaluation is always the same

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
bm: Odometry object owns the robot dimensions

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
cp: add some comments

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
client: send request method updated

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
client: send requested position

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
client: send position request

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
bm: wheel speed check

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
build: build.sh renamed to run

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
client: comments

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
comments: add missing comments

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
build: generate documentation on compilation

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
doc: update doxigen config file

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
client: limit the receive time properly

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
status: protect multiple include

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: correct documentation

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
controlSW: code standard fixed

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
bm: transfer client to separate location

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
stopwatch: limit maximal memroy usage

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: correct the time interval

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

odometry: correct the time interval

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
stopwatch: check index on request

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

MichalDobis commented on src/blackmetal/launch/blackmetal.launch.py:
```diff
@@ -22,12 +22,11 @@ def generate_launch_description():
-			parameters = [config_log],
-			emulate_tty=True
-		)
+		# Node(
+		# 	package='blackmetal',
+		# 	executable='log_server',
+		# 	output='screen',
+		# 	emulate_tty=True
+		# )
```
preco zakomentovane?

Fildo7525 replied:
Hej to som zabudol odkomentovat ale je to v komite DO-NOT-MERGE

Fildo7525 replied:
odkomentovane

MichalDobis commented on src/blackmetal/src/communication/controlSW/BlackMetal.cpp:
```diff
@@ -3,48 +3,65 @@
+	return m_odometry->getWheelRadius();
+}
+
+void BlackMetal::checkWheelSpeed(double &wheelSpeed)
```
toto robi check? A zmeni hodnotu argumentu? Ak je to check, tak by to malo vraciat bool. Ak to ma nieco modifikovat, tak by sa to malo volat modifyWheelSpeed. Alebo podla toho co to robi, asi by som to nazval clampWheelSpeed. Alebo jednoducho pouzi funkciu std::clamp :) https://en.cppreference.com/w/cpp/algorithm/clamp

Fildo7525 replied:
jo opravene

MichalDobis commented on src/blackmetal/src/communication/controlSW/Odometry.cpp:
```diff
@@ -0,0 +1,185 @@
+		auto rws_start = jsonMessage.find_last_of(':') + 1;
+		auto rws_end = jsonMessage.find_last_of('}');
+		rws = std::stol(jsonMessage.substr(rws_start, rws_end));
+	} catch (std::out_of_range &e) {
+		DBG("Attempting to receive the json on next read, current: " << std::quoted(jsonMessage));
+		std::string nextAttempt;
+		nextAttempt = m_controlClient->robotVelocity();
+		return obtainWheelSpeeds(std::move(nextAttempt));
+	} catch (std::exception &e) {
+		ERR(e.what() << " with string " << std::quoted(jsonMessage));
+		return {0,0};
+	}
```
toto sa aj zvykne stavat?

Fildo7525 replied:
Mam tam toho ze chcem zistit co to tam je potrebne. Ale momentalne uz asi nie, lebo priamo sa to triedi v klientovi. Dam to potom prec

Fildo7525 replied:
Uvazil som ze to tam nebude treba. Do tejto funkcie sa dostavaju len mesage ktore boli kliento m dane do rady.

MichalDobis commented on src/blackmetal/src/communication/controlSW/Odometry.hpp:
```diff
@@ -0,0 +1,147 @@
+
+#include <memory>
+#include <thread>
+class Odometry;
```
Preco tato deklaracia?

Fildo7525 replied:
Musi to tam byt lebo tam je cyklicka dependencia. Bez tejto deklaracie by to neslo skompilovat. Teda aspon tam bola ked som to tam pisal

Fildo7525 replied:
som sa trocha pohras includmi a uz to tam nemusi byt

MichalDobis reviewed (approved):
Super :) Komentare iba na zopar detailov

Commit:
bm: use std::clamp instead of manual check

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: no need to declare header of BlackMetal class

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
odometry: the message is ok in this function

When we use the ts::Queue the client while receiving sorts out the checking
messages. Thus to this function are passed only arguments from
m_odometryMessages queue.

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
log: we do not need mergeLog. We log to run-***/all.log

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>

Commit:
coments: added some coments

Signed-off-by: Filip Lobpreis <f.lobpreis@gmail.com>
